File: go_src/3.go
func lengthOfLongestSubstring(s string) int {
    n := len(s)
    if n == 0 {return 0}
    m := make(map[byte]int)
    maxLen := 0
    for left, right := 0, 0; right < n; right++ {
        if p, ok := m[s[right]]; ok {
            left = int(math.Max(float64(left), float64(p)))
        }
        m[s[right]] = right + 1
        maxLen = int(math.Max(float64(maxLen), float64(right - left + 1)))
    }
    return maxLen
}

File: java_src/3.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }

        int left = 0, right = 0;
        int maxLen = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        for (; right < s.length(); right++) {
            if (map.containsKey(s.charAt(right))) {
                left = Math.max(map.get(s.charAt(right)), left);
            }
            maxLen = Math.max(right - left + 1, maxLen);
            map.put(s.charAt(right), right + 1);
        }
        return maxLen;
    }
}
